mabwiser/treebandit.py

```diff
@@ -0,0 +1,173 @@
+# -*- coding: utf-8 -*-
+# SPDX-License-Identifier: Apache-2.0
+
+from copy import deepcopy
+from typing import Union, Dict, List, NoReturn, Optional, Callable
+
+import numpy as np
+from sklearn.tree import DecisionTreeClassifier
+
+from mabwiser.base_mab import BaseMAB
+from mabwiser.greedy import _EpsilonGreedy
+from mabwiser.linear import _Linear
+from mabwiser.popularity import _Popularity
+from mabwiser.rand import _Random
+from mabwiser.softmax import _Softmax
+from mabwiser.thompson import _ThompsonSampling
+from mabwiser.ucb import _UCB1
+from mabwiser.utils import reset, argmax, Arm, Num, _BaseRNG
+
+
+class _TreeBandit(BaseMAB):
+    def __init__(self, rng: _BaseRNG, arms: List[Arm], n_jobs: int, backend: Optional[str],
+                 lp: Union[_EpsilonGreedy, _Linear, _Popularity, _Random, _Softmax, _ThompsonSampling, _UCB1],
+                 tree_parameters: Dict):
+        super().__init__(rng, arms, n_jobs, backend)
+        self.lp = lp
+        self.tree_parameters = tree_parameters
+        self.tree_parameters["random_state"] = rng.seed
+
+        self.arm_to_tree = None
+        self.arm_to_rewards = None
+
+        # Initialize the arm expectations to nan
+        # When there are neighbors, in the leaf node of the tree, expectations of
+        # the underlying learning policy is used
+        # When there are no neighbors, return nan expectations
+        reset(self.arm_to_expectation, np.nan)
+
+    def fit(self, decisions: np.ndarray, rewards: np.ndarray, contexts: np.ndarray = None) -> NoReturn:
+
+        # Reset the decision trees of each arm
+        self.arm_to_tree = {arm: DecisionTreeClassifier(**self.tree_parameters) for arm in self.arms}
```

Comment:
So for every arm without training data there will be an unfitted DecisionTreeClassifier? Seems like it would be good to track those unfitted arms here as an attribute of the class (self.unfitted_arms) for later use